### ISSUE TITLE
Some guidance on getting this working

### ISSUE THREAD

**Issue by Analect** [dgonzalez/seneca-mesh-kubernetes#1]
@paolochiodi @dgonzalez 
Thanks for assembling this demo.  I have been trying to get this basic demo working in a 1.5.2 cluster on GKE, ahead of trying to replicate it for my own app, but without success.

There are no manifest files posted in here, so I assembled a fairly basic `docker-compose.yml` ...
```
version: '2'
services:
  seneca-base:
    image: myregistry/seneca/seneca-base:0.1
    ports:
      - "39000:39000"
  seneca-act:
    image: myregistry/seneca/seneca-act:0.1
    ports:
      - "3000:3000"
    depends_on:
      - base
```
... and then used `kompose --file docker-compose.yml convert --chart` go generate a helm chart.

The generated manifest files combined together look like this.  I ended up having to add `app: seneca-act` and `app: seneca-base` labels manually.

```
apiVersion: v1
kind: Service
metadata:
  creationTimestamp: null
  labels:
    service: seneca-base
    app: seneca-base
  name: seneca-base
spec:
  ports:
  - name: "39000"
    port: 39000
    protocol: TCP
    targetPort: 39000
  selector:
    service: seneca-base
status:
  loadBalancer: {}
---
apiVersion: v1
kind: Service
metadata:
  creationTimestamp: null
  labels:
    service: seneca-act
    app: seneca-act
  name: seneca-act
spec:
  ports:
  - name: "3000"
    port: 3000
    protocol: TCP
    targetPort: 3000
  selector:
    service: seneca-act
status:
  loadBalancer: {}
---
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  creationTimestamp: null
  name: seneca-base
spec:
  replicas: 1
  strategy: {}
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: seneca-base
    spec:
      containers:
      - image: myregistry/seneca/seneca-base:0.1
        name: seneca-base
        ports:
        - containerPort: 39000
          protocol: TCP
        resources: {}
      restartPolicy: Always
status: {}
---
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  creationTimestamp: null
  name: seneca-act
spec:
  replicas: 1
  strategy: {}
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: seneca-act
    spec:
      containers:
      - image: myregistry/seneca/seneca-act:0.1
        name: seneca-act
        ports:
        - containerPort: 3000
          protocol: TCP
        resources: {}
      restartPolicy: Always
status: {}
```
Are there any glaring mistakes in there?  The reason being that on deployment, the services are terminating immediately.  I'm wondering if perhaps the kubernetes API that the plugin references may have changed.

![image](https://cloud.githubusercontent.com/assets/4063815/23138396/2eaa22b2-f79f-11e6-8197-025bfee58740.png)

Also, I notice that port `39000` is being used here for base, as against the default `39999`.  Perhaps that's a factor?

In layman's terms, how is seneca-mesh leverage networking within kubernetes to function?  Presumably it's possible to scale up certain instances of services and it just works.  Thanks for any guidance you might have to get this to work.

**Comment by Analect** [comment]
Sorry ... I clearly hadn't exhausted the possibilities.  When I started the above manifests manually ... base-deployment, base-service, act-deployment, act-service, I think that gave the base the time it needed to be running ahead of act starting.

![image](https://cloud.githubusercontent.com/assets/4063815/23139029/68a2cba6-f7a2-11e6-89ab-9ea627a5ea10.png)

![image](https://cloud.githubusercontent.com/assets/4063815/23139043/77ff9f0c-f7a2-11e6-969e-f81c438343e2.png)

I would still value a better understanding of mesh working in the context of kubernetes networking, if you get a moment. Thanks.

**Comment by paolochiodi** [comment]
Hi @Analect thank for checking this out.

There are two problems in using senecas-mesh on kubernetes:
- kubernetes networking doesn't allow to send info on broadcast address
- inside containers node servers usually listen to 127.0.0.0 instead of the "public" address

This demo solves the two problems by leveraging kubernetes api to fetch ip addresses for current container and other bases containers and using those instead of the default/autodiscovery of seneca-mesh.
This means that at least a base must be started before other services.

# Verbose explanation
SWIM, the protocol on which `seneca-mesh` is based work by briefly advertising a new node on the broadcast address of the local network, so that other live nodes can detect it, add it to the mesh and start sharing with the new node the architecture of existing mesh.
`seneca-mesh` nodes advertise themselves sending the ip address on which they are listening.

Unfortunately kubernetes doesn't support sending info on broadcast. 
Another problem is that when you create a nodejs instances in a container you typically listen to 127.0.0.1 or localhost and leave to kubernetes to handle the routing.

However, this means that a new `seneca-mesh` node advertise itself as living on 127.0.0.1 and other nodes that receive the message won't be able to find it, because kubernetes isolate the containers and the 127.0.0.1 they can see is different from the one the new node is actually listening on.

In order to make everything work we need to listen on a different ip address, the one visible from the outside, and also fetch the ip addresses for the bases - so we know how to reach those.

`seneca-kubernetes` helps in this by pre-loading some data about existing containers in the network and in the demo we use that data to get the public ip of the container https://github.com/dgonzalez/seneca-mesh-kubernetes/blob/master/seneca-act/act.js#L43 and the ip of other bases https://github.com/dgonzalez/seneca-mesh-kubernetes/blob/master/seneca-act/act.js#L30. 
Please not that we detect bases using a label that by convention is `seneca-base` https://github.com/dgonzalez/seneca-mesh-kubernetes/blob/master/seneca-act/act.js#L14

I'm not sure if I've been able to explain myself cleary - let me know if I need to elaborate more

**Comment by Analect** [comment]
@paolochiodi 
Thanks for the quick response and explanation.

Would you say that it's suboptimal to be running seneca-mesh in a kubernetes context.  If services come and go and if they (the services) are hampered in terms of discovering eachother then does this operation to "fetch the ip addresses for the bases" have to be an ongoing process? .... and does one have to assign fixed ports to each service?

I see from [here](https://github.com/senecajs/seneca-mesh/issues/48#issuecomment-256782439) that someone else using seneca-mesh in a kubernetes context needed to maintain a register of bases on consul.  Have you seen this implemented elsewhere?

Are the short-comings in kubernetes likely to get resolved ... in terms of new features / roadmap?  Without having any expertise in this area, does the built-in [DNS service](https://kubernetes.io/docs/admin/dns/) offer an alternative more robust way for seneca services to communicate with eachother? 

By chance, a few days ago, I was searching for sidecar-related aspects in kubernetes ... and came across this separate [sidecar](http://relistan.com/sidecar-service-discovery-for-all-docker-environments/) solution form newrelic ... which seems to implement the SWIM protocol in a kubernetes context ... I'm not clued-in enough to determine if what they are doing is any different to your solution.  I do notice that the main [repo](https://github.com/newrelic/sidecar) doesn't mention much about operating in a kubernetes context as among its options though.

**Comment by paolochiodi** [comment]
SWIM protocol and by extension seneca-mesh doesn't need to continually update the addresses of the bases.
Actually, the concept of "base" doesn't exist in the SWIM protocol and has been introduced by seneca-mesh for simplification.

For a new node to join a SWIM network it only needs the address of another node: once it connect to it it will (eventually) receive the addresses of all other nodes in the mesh.

The concept of bases has been introduced in `seneca-mesh` to somehow design a service as a stable base for the mesh or, put in another way, the service that need to be started first.

Once the mesh is up and running every new node can join by knowing another node of the mesh. New "bases" can be introduced and as long as they connect to one of the existing nodes and then they will be part of the mesh (and could be used as bases for other nodes to join).

To try and answer you questions:
- "fetch the ip addresses for the bases" is not an ongoing process: once the node has joined the mesh it will receive the updates on the mesh architecture automatically. But every new node will need to fetch the ip of the bases during startup (or, even better, a list of known nodes of the mesh. One can argue that we can consider every node in a kubernetes mesh as good to be bases)
- is it sub-optimal? I think it has both upsides and downsides. Surely kubernetes offers by default another way to solve service discovery (internal dns, services...) that may be worth checking. `seneca-mesh` "magical" autodiscovery is a bit limited in non ideal environments and will need a bit of configuration to work properly.
- consul is just another way to find out ip addresses for the bases. It is officially supported by `seneca-mesh`, but personally I didn't see it implemented (if I recall properly @dgonzalez briefly tested it for this demo). The overall concept doesn't change: if broadcasting isn't available we need to know the address of one of the nodes already in the mesh. For that you can use kubernetes api, consul, environment variables or any other method
- for roadmaps regarding seneca and seneca-mesh you should check with @rjrodger. From what I can tell I don't think there will be changes seneca-mesh to better support kubernetes, not for a lack of will but because the solution in this demo (or using consul) are already the best solution we can achieve. And yes, kubernetes dns services are in my opinion a good way to have the micro services communicate each other (as long as you use the http or tcp transport)
- sidecar: I don't have any knowledge on this so can't offer proper advice on it. From a brief reading of the post you linked it seems to me that they incorporated a service registry (like consul) inside the solution and thus solved the problem of finding the "bases"

**Comment by Analect** [comment]
@paolochiodi .  That's very helpful. Appreciate your time explaining this.  I'll close this out.

**Comment by paolochiodi** [comment]
👍
